packages/babel-plugin-transform-modules-systemjs/test/fixtures/systemjs/module-level-variable-destructuring/output.mjs

```diff
@@ -0,0 +1,23 @@
+System.register(["./x.js"], function (_export, _context) {
+  "use strict";
+
+  var x, _x;
+
+  return {
+    setters: [function (_xJs) {
+      x = _xJs.x;
+    }],
+    execute: function () {
+      if (true) {
+        ({
+          x: _x
```

Comment:
Why `const { x } = { x: 1 }` is also transformed? I would expect `_x` is registered in the same scope of the local `x` declaration:

```js
if (true) {
  const { x: _x } = { x: 1 };
  console.log(_x);
}
```

Reply:
Hoisted vars are renamed, So I think that's the reason for this.

Reply:
What should we change in this?

Reply:
Oh right, the problem is that we should only hoist `var` declarations. Then we can pass `"var"` instead of `null` to `hoistVariables` in the `systemjs` plugin (or we can just delete that parameter, since `"var"` is the default value).

Reply:
I have been thinking about this, and the proper fix probably is in the `systemjs` plugin itself.

We should not hoist all variables (`var` and `let`/`const`), but only `var` and top-level `let`/`const`. We can do this by:
1. In the big loop over the program body (in the `Program.exit` visitor), we convert all the exported `let`/`const` bindings to `var`. This is safe, because exported `let`/`const` bindings are necessarily defined in the top-level scope and not in nested blocks.
2. When calling `hoistVariables`, we should remove the third parameter (`null`) so that it only hoists `var` and not also `let`/`const`.

Reply:
For (1), be careful because `export let x = 1;` and `export { x }; let x = 1;` are handled in two different code paths and both should still work.

Reply:
Also, it looks like after this change `hositVariables` will only be used to hoist `var` declarations, so in Babel 8 we can remove `let` support :smile:

Reply:
I think the systemjs transform hoists more variables than it should've. Specifically, it is hoisting block scoped variable, which should never affect the exported specifiers because export statement is always on the top level. For example, the expected output of
```js
import { x } from './x.js';

if (true) {
	const { x } = { x: 1 };
	console.log(x);
}

new (class extends x {})();
```
should be
```js
System.register(["./x.js"], function (_export, _context) {
  "use strict";

  var x;
  return {
    setters: [function (_xJs) {
      x = _xJs.x;
    }],
    execute: function () {
      if (true) {
        const {
          x
        } = {
          x: 1
        };
        console.log(x);
      }

      new class extends x {}();
    }
  };
});
```

So I suggest we fix this in the systemjs transform. We can inline the `hoistVariable` visitor, but skip those variable declarations, if they are 1) of kind "let" or "const" and 2) defined in a new scope (other than the top level scope). For more context about why `let` or `const` should be hoisted, see https://github.com/babel/babel/pull/8104.

I don't think we need to rename the bindings if the path is a block-scoped declaration under block statement, because 1) at the hoist-variable-helper level, we don't know where the bindings will be hoisted to, neither do we know they will conflict with bindings in the target scope 2) even if there are conflict bindings, it should be up to the `emitFunction`'s responsibility to ensure the hoisted variables can be legally registered, either by renaming the hoisted variable, or renaming its own bindings. 3) a block-scoped variable can be defined under any BlockParent, including BlockStatement.

Reply:
I think the systemjs transform hoists more variables than it should've. Specifically, it is hoisting block scoped variable, which should never affect the exported specifiers because export statement is always on the top level. For example, the expected output of
```js
import { x } from './x.js';

if (true) {
	const { x } = { x: 1 };
	console.log(x);
}

new (class extends x {})();
```
should be
```js
System.register(["./x.js"], function (_export, _context) {
  "use strict";

  var x;
  return {
    setters: [function (_xJs) {
      x = _xJs.x;
    }],
    execute: function () {
      if (true) {
        const {
          x
        } = {
          x: 1
        };
        console.log(x);
      }

      new class extends x {}();
    }
  };
});
```

So I suggest we fix this in the systemjs transform. We can inline the `hoistVariable` visitor, but skip those variable declarations, if they are 1) of kind "let" or "const" and 2) defined in a new scope (other than the top level scope). For more context about why `let` or `const` should be hoisted, see https://github.com/babel/babel/pull/8104.

I don't think we need to rename the bindings if the path is a block-scoped declaration under block statement, because 1) at the hoist-variable-helper level, we don't know where the bindings will be hoisted to, neither do we know they will conflict with bindings in the target scope 2) even if there are conflict bindings, it should be up to the `emitFunction`'s responsibility to ensure the hoisted variables can be legally registered, either by renaming the hoisted variable, or renaming its own bindings. 3) a block-scoped variable can be defined under any BlockParent, including BlockStatement.

Reply:
Okay I understand, I will wait for @nicolo-ribaudo re-review before working on.

Reply:
I dropped my review to re-review _after_ the updates 😛

Reply:
Oh sorry, I will start working on. 😂

Reply:
If third parameter is removed from the `hoistVariable` it removes all the `_export` calls from the outputs, and doesn't transform the input?

Reply:
Yeah we will still need to hoist these export bindings _and_ set up the metadata from which the `_export` calls are built. I suggest we copy the code from the `hoistVariable` to the systemjs transform, and implement the custom hoisting logic mentioned in https://github.com/babel/babel/pull/14057#discussion_r821846879. Or we can add a new option in the helper, though I lean to the first option since the hoisting logic here is not reused by other plugins.

Reply:
I'm surprised that the current helper doesn't work, because we only need `_export` for things that become `var` in step 1 :thinking: 
Do you mind pushing the current (non working) code you have?

Reply:
I don't have a working branch. We can infer the behaviour of `hoistVariable` helper from the implementation:

https://github.com/babel/babel/blob/dacaab1071a509f334c9c7355c38dd829cda9935/packages/babel-helper-hoist-variables/src/index.ts#L23-L32

The third variable `kind` controls two behaviours: 1) Whether we should check a BlockParent to see if it contains to-be-hoisted variables, and 2) whether we should hoist the `let`-kind variable declarations.

In systemjs transform, what we want to achieve is to hoist `var` variables within block parent (so `kind` has to be `"let"`), but leave `let`-kind declarations within block parent untouched.

Reply:
@The-x-Theorist I hope you don't mind, I pushed two commits to your branch.

[`14e9cfd` (#14057)](https://github.com/babel/babel/pull/14057/commits/14e9cfd1e2183e882e266ed8206680607bf1b080) is how I was thinking to fix the problem, as described in https://github.com/babel/babel/pull/14057#discussion_r821846879:
1. It changes all the top-level `let`/`const` to `var`
2. It only hoists `var`s

The updated tests look good, but it probably needs a few new tests to make sure that both `const x = 1; export { x }` and `export const x  = 1;` work (unless we already have these tests).